src/Streams/StreamWriter.h

```diff
@@ -36,6 +39,19 @@ class StreamWriter {
 		WriteImplementation(vector.data(), vector.size() * sizeof(T));
 	}
 
+	// std::string prefixed by the string's size. The type of integer representing the size must be provided.
+	// Does not null terminate string unless null terminator is specifcally included in passed std::string. 
+	template<typename SizeType>
+	void Write(const std::string& string) {
+		auto stringSize = string.size();
+		if (stringSize > std::numeric_limits<SizeType>::max()) {
+			throw std::runtime_error("String's size is too large to write in provided size field");
+		}
+		auto typedSize = static_cast<SizeType>(stringSize);
+		Write(typedSize);
+		Write(string.data(), string.size());
```

Comment:
There is some inconsistency between the use of `string.data()` and `&string[0]` in this change set.

Additionally, it feels slightly weird to `Write(typedSize)`, and then write `string.size()` characters. I almost feel like the same variable should be used in both places. Though the datatype may feel more natural with the code as written.

Reply:
I prefer string.data(). However in StreamReader, I receive the error below when writing the line: 

```
Read(string.data(), stringSize);
```

Error C2664 'void StreamReader::Read(void *,size_t)': cannot convert argument 1 from 'const char *' to 'void *' OP2Utility c:\ . . . \op2mapimager\op2utility\src\streams\streamreader.h 56

So I was forced to write it there as &string[0]

Looking at this a little closer I'm not sure there is any reason to keep the line with the static_cast. It doesn't do anything. instead it could be written as:
```
Write(string.size());
Write(string.data(), string.size());
```
The variable typedSize and the static_cast can just be removed. After we check the numeric_limit, there isn't any reason to cast anything???

Thanks,
Brett

Reply:
Actually, there is reason for the cast, since the `Write` is dependent on the data size. If you got rid of the variable you would still need the cast to write the correct number of bytes. The following may work well:
```cpp
Write(static_cast<SizeType>(stringSize));
Write(string.data(), stringSize);
```

Depending on the compiler and compiler options, using `string.size()` everywhere may result in multiple calls to the `.size()` method. I suspect most compilers won't be smart enough to realize the value is constant and optimize the extra calls away and just use a cached temporary. Optimizing across calls is more difficult than within a method.

----

As for [`.data()`](https://en.cppreference.com/w/cpp/string/basic_string/data), keep in mind the signature was changed over different C++ versions. Before C++17, it was `const`, so inappropriate to pass to read methods (which write into the passed object). For ['operator[]`](https://en.cppreference.com/w/cpp/string/basic_string/operator_at) the constness matches the constness of the object itself. This naturally works where it's supposed to, and gives compilers errors where needed.